packages/shared/src/hooks/usePersistentContext.ts

```diff
@@ -21,24 +21,27 @@ export default function usePersistentContext<T>(
   key: string,
   valueWhenCacheEmpty?: T,
   validValues?: T[],
-): [T, (value: T) => Promise<void>] {
+): [T, (value: T) => Promise<void>, boolean] {
+  const queryKey = [key, valueWhenCacheEmpty];
   const queryClient = useQueryClient();
 
-  const { data } = useQuery<unknown, unknown, T>(
-    [key, valueWhenCacheEmpty],
-    () => getAsyncCache<T>(key, valueWhenCacheEmpty, validValues),
+  const { data, isFetched } = useQuery<unknown, unknown, T>(queryKey, () =>
+    getAsyncCache<T>(key, valueWhenCacheEmpty, validValues),
   );
 
-  const { mutateAsync: updateValue } = useMutation(
+  const { mutateAsync: updateValue } = useMutation<void, unknown, unknown, T>(
     (value: T) => setCache(key, value),
     {
       onMutate: (mutatedData) => {
-        queryClient.setQueryData(key, mutatedData);
-        return mutatedData;
+        const current = data;
+        queryClient.setQueryData(queryKey, mutatedData);
+        return current;
+      },
+      onError: (_, __, rollback) => {
+        queryClient.setQueryData(queryKey, rollback);
       },
-      onError: () => {},
     },
   );
 
-  return [data, updateValue];
+  return [data ?? valueWhenCacheEmpty, updateValue, isFetched];
```

Comment:
I actually think we should not block this.
If we simply return data it should be quick enough?

Reply:
But then again, your fix would not be breaking, just causes it to re-render twice I guess.
Not quite sure why the test fail then, we use this in other places as well.

@idoshamun What do you think about the above?

Reply:
I think it already re-renders twice since the initial value is undefined while fetching the data - then returns the actual or placeholder value.

With that, I can handle the case outside to set a value if undefined. I was also thinking if we should add an initial value parameter like how it used to be at `usePersistentState`

Reply:
`valueWhenCacheEmpty` should be used after we validate that the cache is not presented. In this case, we actually use it as a default value which is quite different. In `usePersistentState`, we have an `initialValue` parameter that is used specifically for the purpose of a default value. So if we want to stick to the new behavior I suggest renaming accordingly. Please make sure it doesn't break any of the existing usages of this hook because it potentially can.
https://github.com/dailydotdev/apps/blob/master/packages/shared/src/hooks/usePersistentState.ts#L4

Reply:
Cool. Let me just then add a parameter for a fallback value for when the data is yet undefined so we can be sure things will still be compatible with existing implementations.